certbot/lock.py

```diff
@@ -108,17 +175,68 @@ def _lock_success(self, fd):
         # the same device and inode, they're the same file.
         return stat1.st_dev == stat2.st_dev and stat1.st_ino == stat2.st_ino
 
-    def __repr__(self):
-        repr_str = '{0}({1}) <'.format(self.__class__.__name__, self._path)
-        if self._fd is None:
-            repr_str += 'released>'
-        else:
-            repr_str += 'acquired>'
-        return repr_str
-
     def release(self):
+        # type: () -> None
         """Remove, close, and release the lock file."""
+        # It is important the lock file is removed before it's released,
+        # otherwise:
+        #
+        # process A: open lock file
+        # process B: release lock file
+        # process A: lock file
+        # process A: check device and inode
+        # process B: delete file
+        # process C: open and lock a different file at the same path
+        try:
+            os.remove(self._path)
+        finally:
+            # Following check is done to make mypy happy: it ensure that self._fd, marked
+            # as Optional[int] is effectively int to make it compatible with os.close signature.
+            if self._fd is None:  # pragma: no cover
+                raise TypeError('Error, self._fd is None.')
+            try:
+                os.close(self._fd)
+            finally:
+                self._fd = None
+
+
+class _WindowsLockMechanism(_BaseLockMechanism):
+    """
+    A Windows lock file mechanism.
+    On Windows in general, access to a file is exclusive, so opening a file
+    is an effective lock. This default behavior may be modified by an administrator,
+    so we ensure correct behavior with msvcrt. Moreover on Windows a file can be removed
+    only after it has been released: so the concurrency access that may occur on POSIX
+    system is irrelevant here, leading to a quite simple code.
```

Comment:
In fact my knowledge of the specific lock operations on Windows improved significantly since I wrote this comment at the beginning. And it became irrelevant.

I may have thought that an administrator could change the default behavior about exclusive locks on file handlers on a particular reading. But from the further verifications I did, I did not ever find an indication that this could be done in fact.

Finally things are very clear from every documentation I found: Windows kept the behavior that has been designed since MS-DOS. By default, the low level APIs on Windows about file handler give an exclusive lock on the file to the caller. And theses APIs allow to acquire on demand a file handler in share mode, that makes possible to obtain several file handlers to the file from different processes.

Many years ago, in the majority of the situation where a process was opening a file would lead to a exclusive lock, and other processes should wait to take a file handler on an already open file, or an exception would be raised.

From my experience, I can not found now any software that would open a file in exclusive mode. It seems that progressively every software made changes to always open files in share mode. And I can read and write files on Windows from different processes in parallel. 

Python is not an exception here. Calling open, os.open, or os.fdopen never create an exclusive lock on the opened file. So for that matter, it appears that the behavior of Python is consistent between Linux and Windows.

So you need an explicit lock on Windows. And this is done using the specific logic that allows to lock a part of the bytes on a file, through msvcrt. So yes, msvcrt logic is mandatory here to have a real lock.

Reply:
Final note: even if you acquire a file handler I share mode, on Windows, you still cannot delete a file before every file handler opened has been closed. This makes impossible to have the race condition observed on Linux.

Reply:
Holy hell. What a rabbit hole that was of reading Windows API docs (or lack there of) and Python source code.

I agree we need this. There are methods in Windows that allow you to create a file that can be deleted by another process such as [CreateFileA](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-createfilea) and [people have proposed making delete sharing the default in Python](https://bugs.python.org/issue15244), but that has not happened.

I still struggle to find definitive documentation about the sharing behavior here so if you have found it, please send me a link, but it does seem like both we need `msvcrt` and the remove call is OK.

Reply:
I really wish I could find more definitive documentation about all this. It'd go a long way towards making me feel comfortable that we're doing the "Right Thing"™.

With that said, we've observed the need for `msvcrt` and core Python developers talking about how deleting an open file on Windows isn't possible with Python's current APIs so I think we're probably safe here.

Can you update this comment to reflect what we've learned? For instance "On Windows in general, access to a file is exclusive, so opening a file is an effective lock." doesn't appear to be accurate anymore.

Reply:
Ah yes. I forgot to update this conversation. I have found documentation that explains what happen exactly when delete share flag is enabled on a file handler.

I will do it tomorrow, but to sum up, it is definitely not the same than in Linux, and put the file in a twilight zone as soon as one process ask for deletion, making all current files handler unusable, throw exception if you try to acquire one more on it, and delete the file when all file handlers are released.

The file system stay definitively in a defensive approach to first release a file resource, before deleting it. Still nothing about an action that could change this. 

I will update here, and in the comments, accordingly.

Reply:
So, a discussion on MSDN explaining what is happening to a file with `FILE_SHARE_DELETE`: https://blogs.msdn.microsoft.com/oldnewthing/20071109-00/?p=24553

Reply:
Again, this functionalities require nonetheless to manipulate explicitly low level methods: http://blog.httrack.com/blog/2013/10/05/creating-deletable-and-movable-files-on-windows/

Reply:
Thanks for the links.

To summarize here, we have observed the need for `msvcrt` locking because we have seen Python allow us to open the same file in multiple processes on Windows.

For deleting a file opened by another process, I'd love to see a simple statement in Microsoft documentation like:

> It is impossible to delete a file opened by another process across all versions of Windows unless X, Y, or Z.

I don't think such a statement exists, however, we have observed that Python does not allow us to do that with our current code and dependencies and we've seen Python developers talking about how:

> On Windows this is not possible for files opened using io.open() or os.open().

Given that, I think this code is correct with the current Python and Windows APIs.

@adferrand, can you fix up the comment in the code here and resolve merge conflicts?

Reply:
Done!